### ISSUE TITLE
trained on vitonhd for 130 epochs but result still seems to be weak, is it normal?

### ISSUE THREAD

**Issue by xdfeng07** [Gal4way/TPD#4]
Currently at 130 epoch, the model could only some how determine that the repainting area should be around the upper body of the person, but still with small percentage of failure and put the repainting area on the background. and for the consistency of the input cloth and tryon result, it didn't succeed even once during my trials.

Is it normal? or do I get something wrong with the training?

**Comment by xdfeng07** [comment]
and the model at 130 epoch always failed at getting the repainting area correct, always mistake the long/short sleeves, and the neck area also got wrong usually.

**Comment by Gal4way** [comment]
could u show your results here and i can help u fix it.

and present your training commands and cfg

**Comment by xdfeng07** [comment]
sure,
this is my command line to run training:
CUDA_VISIBLE_DEVICES=1,4,5,6,7 python main.py \
--logdir train_logs/VITONHD/ \
--pretrained_model model_prepare_tpd.ckpt \
--base tpd_configs/train/train_VITONHD.yaml \
--scale_lr False \
--name VITONHD_release_person_combine_garment_240epochs

the config is as follows:
model:
  base_learning_rate: 1.0e-05
  target: ldm.models.diffusion.ddpm.LatentDiffusion
  params:
    linear_start: 0.00085
    linear_end: 0.0120
    num_timesteps_cond: 1
    log_every_t: 200
    timesteps: 1000
    first_stage_key: "inpaint"
    cond_stage_key: "image"
    image_size: 64
    channels: 5
    cond_stage_trainable: true   # Note: different from the one we trained before
    conditioning_key: crossattn
    monitor: train/loss
    u_cond_percent: 0.2
    scale_factor: 0.18215
    use_ema: False

    scheduler_config: # 10000 warmup steps
      target: ldm.lr_scheduler.LambdaLinearScheduler
      params:
        warm_up_steps: [ 10000 ]
        cycle_lengths: [ 10000000000000 ] # incredibly large number to prevent corner cases
        f_start: [ 1e-6 ]
        f_max: [ 1. ]
        f_min: [ 1. ]

    unet_config:
      target: ldm.modules.diffusionmodules.openaimodel.UNetModel
      params:
        image_size: 32 # unused
        in_channels: 18
        out_channels: 5
        model_channels: 320
        attention_resolutions: [ 4, 2, 1 ]
        num_res_blocks: 2
        channel_mult: [ 1, 2, 4, 4 ]
        num_heads: 8
        use_spatial_transformer: True
        transformer_depth: 1
        context_dim: [768]
        use_checkpoint: True
        legacy: False
        add_conv_in_front_of_unet: False

    first_stage_config:
      target: ldm.models.autoencoder.AutoencoderKL
      params:
        embed_dim: 4
        monitor: val/rec_loss
        ddconfig:
          double_z: true
          z_channels: 4
          resolution: 256
          in_channels: 3
          out_ch: 3
          ch: 128
          ch_mult:
          - 1
          - 2
          - 4
          - 4
          num_res_blocks: 2
          attn_resolutions: []
          dropout: 0.0
        lossconfig:
          target: torch.nn.Identity

data:
  target: main.DataModuleFromConfig
  params:
      batch_size: 10
      wrap: False
      train:
          target: ldm.data.dataset_VITONHD.try_on_dataset_VITONHD
          params:
              state: train
              dataset_dir: path_to_vitonhd
              order: paired
              image_size: 224

lightning:
  trainer:
    max_epochs: 240
    num_nodes: 1
    accelerator: 'gpu'
    gpus: "0,1,2,3,4"
    limit_val_batches: 0.5
    check_val_every_n_epoch: 500

the inference result is like:
![image](https://github.com/Gal4way/TPD/assets/20915534/0d667e11-dadb-406c-8094-f30a5163df03)
![image](https://github.com/Gal4way/TPD/assets/20915534/272c10ee-3035-4e52-b94e-b3fda1d95e52)

**Comment by Gal4way** [comment]
1. We trained the pre-trained model using 4 GPUs with 10 batches per GPU, resulting in a total of 240 training epochs. Given that you are utilizing 5 GPUs in your setting, which means a larger batch size, 130 epochs may not be sufficient. I would suggest training your model for a greater number of epochs, potentially around 300 or 350.

2. Make sure your reference garment image is correctly sent to the model and concatenated with the source person image. In my experience, it doesn't take long for the model to learn how to transfer the reference garment to the maked person image.

2. In the wild virtual try-on is quite challenging for current methods. One of the key issues could stem from the dataset itself. If you have reviewed the images in the dataset, you'll notice a distinct difference between those and images captured in the wild. Therefore, I recommend validating your model's effectiveness with images from the VITON-HD dataset. Additionally, consider collecting an in-the-wild training set and enhance the model's performance in such scenarios.

I hope these suggestions are helpful to u.

**Comment by Gal4way** [comment]
> Hey, I think it's not working on dynamic height and width, I mean I tried inferencing on 768x1024 but it generates the same 384x512 output.
> 
> Can we train it on high resolution using this training script?

Yes, you can. You will need to modify the dataset.py file and the inference/training scripts to ensure that the model utilizes the desired resolution.

**Comment by Gal4way** [comment]
> Tried training it on high resolution. Just to confirm; we are concatenating garment image with each image i.e. with person, with pose n all, right? Also, it's taking 55GB with bs of 1 on high res.

Yes, that's exactly what we did in our experiments. Increasing the resolution indeed leads to higher GPU memory usage. Therefore, a more common approach in the industry is to first generate try-on results at a resolution of 512x384 (or lower), and then upscale to the desired resolution using a super-resolution model. We hope this helps. If you find our method useful, feel free to leave us a star; we would really appreciate it.
